File: day08/day08a.py
#!/usr/bin/python3

with open('day08.txt', 'r') as f:
    # Parse each line into operation and argument, with array to store whether each command has been executed
    lines = f.readlines()
    ops = [line[:3] for line in lines]
    args = [int(line[4:]) for line in lines]
    executed = [False for line in lines]

    # Keep track of a program counter and accumulator
    pc = 0
    acc = 0

    # Keep executing commands, stopping before any instruction is repeated
    while not executed[pc]:
        executed[pc] = True
        command = ops[pc]
        if command == 'nop':
            pc += 1
        elif command == 'acc':
            acc += args[pc]
            pc += 1
        elif command == 'jmp':
            pc += args[pc]

    print(acc)

File: day08/day08a.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

int main() {
  // Read input from file
  std::ifstream file("day08.txt");
  std::string line;
  std::vector<std::string> ops;
  std::vector<int> args;
  std::vector<bool> executed;

  // Parse each line into operation and argument, with array to store whether each command has been executed
  int arg = 0;
  while(std::getline(file, line)) {
    ops.push_back(line.substr(0, 3));

    std::stringstream arg_ss(line.substr(4, std::string::npos));
    arg_ss >> arg;
    args.push_back(arg);

    executed.push_back(false);
  }

  // Keep track of program counter and accumulator
  int pc = 0;
  int acc = 0;

  // Keep executing commands, stopping before any instruction is repeated
  while(!executed[pc]) {
    executed[pc] = true;
    std::string &command = ops[pc];

    if(command == std::string("nop")) {
      pc++;
    } else if(command == std::string("acc")) {
      acc += args[pc];
      pc++;
    } else if(command == std::string("jmp")) {
      pc += args[pc];
    }
  }

  std::cout << acc << std::endl;
  return 0;
}
